test/native_objc_test/block_test.dart

```diff
@@ -0,0 +1,73 @@
+// Copyright (c) 2022, the Dart project authors. Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+// Objective C support is only available on mac.
+@TestOn('mac-os')
+
+import 'dart:ffi';
+import 'dart:io';
+
+import 'package:ffigen/ffigen.dart';
+import 'package:path/path.dart' as path;
+import 'package:test/test.dart';
+import 'package:yaml/yaml.dart';
+import '../test_utils.dart';
+import 'block_bindings.dart';
+
+void main() {
+  late BlockTestObjCLibrary lib;
+
+  group('Blocks', () {
+    setUpAll(() {
+      logWarnings();
+      final dylib = File('test/native_objc_test/block_test.dylib');
+      verifySetupFile(dylib);
+      lib = BlockTestObjCLibrary(DynamicLibrary.open(dylib.absolute.path));
+    });
+
+    test('generate_bindings', () {
+      final config = Config.fromYaml(loadYaml(
+          File(path.join('test', 'native_objc_test', 'block_config.yaml'))
+              .readAsStringSync()) as YamlMap);
+      final library = parse(config);
+      final file = File(
+        path.join('test', 'debug_generated', 'block_test.dart'),
+      );
+      library.generateFile(file);
+
+      try {
+        final actual = file.readAsStringSync();
+        final expected = File(path.join(config.output)).readAsStringSync();
+        expect(actual, expected);
+        if (file.existsSync()) {
+          file.delete();
+        }
+      } catch (e) {
+        print('Failed test: Debug generated file: ${file.absolute.path}');
+        rethrow;
+      }
+    });
+
+    test('BlockTester is working', () {
+      // This doesn't test any Block functionality, just that the BlockTester
+      // itself is working correctly.
+      final blockTester = BlockTester.makeFromMultiplier(lib, 10);
+      expect(blockTester.call(123), 1230);
+      final intBlock = blockTester.getBlock();
+      final blockTester2 = BlockTester.makeFromBlock(lib, intBlock);
+      expect(blockTester2.call(456), 4560);
+    });
+
+    test('Block from function pointer', () {
+      final block = ObjCBlock.fromFunctionPointer(
```

Comment:
Is this API meant for end-users? Because this seems like a pretty dangerous (in that it is not type safe) and difficult to use.

Reply:
`block.pointer` is an escape hatch like `castFrom`, and I'm hoping to eliminate all the cases where it's necessary. The rest of the API is type safe afaik.

There's a bit of a limitation to how friendly we can make the API, because `Pointer.fromFunction` is restricted in how it can be used. There's static analysis that turns the function passed to it into a C function pointer, and that analysis requires that there is no intermediate stage in between that function and the `Pointer.fromFunction` call. In other words, it's always going to be the case that the user has to call Pointer.fromFunction themselves.

But if you have a specific suggestion about how to make it more ergonomic, I'd love to hear it.

Reply:
Please help me clarify for myself what we're trying to do here. The mental model that I have for C APIs is that ffigen generates a low level API (with lots of `Pointer`s and only top level functions), and that a user manually wraps that in an object-oriented API. The object-oriented API does not expose any `dart:ffi` classes, only normal Dart classes. (See https://www.youtube.com/watch?v=2MMK7YoFgaA for the mental model.)

Now, because Objective C is already object-oriented, we can immediately generate an object-oriented API. `class BlockTester` does not extend or implement a `dart:ffi` type.

When using a low level C binding without the object-oriented API, a user would just use `Pointer.fromFunction(_add100, 999)`.
In a high level object-oriented API where the API owner owns the callback function. The API owner would just inline the `fromFunction` in the implementation behind the API.
However, in a high level object-oriented API where the API user provides arbitrary callbacks this is not possible.

The typical workaround is to not use `fromFunction` on the users' top-level functions, but create a `Map<int, RetType Function(ArgumentTypes)>` in Dart, store the users' closures there, and use `fromFunction` on a function defined within the API. However, this requires an extra integer argument to be passed to the native library, and the native library should pass that integer back in its callback. This can of course be added if the native library is owned by the person making the Dart API. However, if the native library is not owned by the person writing the object-oriented API, this workaround doesn't work.

I don't see a good workaround for the user having to do `fromFunction` themselves here. The users doing `fromFunction` here is relatively type safe because the `Pointer` has type arguments with the native function signature.

FYI: `fromFunction` needs to statically know its target to be able to do AOT compilation.

I don't see a way for `fromFunction` supporting closures unless users specify how many extra trampoline entries need to be generated, and a hard crash if someone wants one more closure:

* https://github.com/dart-lang/sdk/issues/48865

> block.pointer is an escape hatch like castFrom, and I'm hoping to eliminate all the cases where it's necessary. 

I agree, those cases should be eliminated. The `block` is the objectoriented API, and the `block.pointer` leaks the abstraction by showing the underlying `dart:ffi` object. Can we eliminate it in this PR?

Reply:
I have the following suggestion: unlike naked function pointers, blocks _are_ essentially closures. 

You are already using this (because you are storing `target` separately from the `invoke` trampoline).

You can take this one step further and allow wrapping any Dart closure (not just an FFI pointer to a static function) into a block. Instead of storing target of the call as a function pointer, you can put closure into a persistent handle and store that handle into the block object. Then `invoke` just unwraps the handle and trampolines to the target.

If dealing with persistent handles is too wonky (it requires calling methods of DL API through FFI) you can just setup your own array on the side and use indices into that array instead of handles. 

The only complication here is that we would need to start properly disposing blocks allocated on Dart side to avoid resource leaks and we would need to provide `copy` and `dispose` callbacks that do a bit of reference counting on "closure handles/ids".

But other than that this should work.

Reply:
Yeah, I'm planning to support converting closures to blocks as a follow up PR. When you say "blocks are essentially closures", are you suggesting that I shouldn't support the function pointer case at all?

Reply:
I think it is fine to support both pointers and full blown closures. 

I was mostly reacting to @dcharkes's comment (and associated feature request https://github.com/dart-lang/sdk/issues/48865 ) --- trying to highlight that https://github.com/dart-lang/sdk/issues/48865 is not actually needed for Obj-C blocks.